src/middlewared/middlewared/plugins/vm.py

```diff
@@ -1284,13 +1292,16 @@ def do_delete(self, id, data):
                 )[-1]
                 self.middleware.call_sync('zfs.dataset.delete', disk_name)
 
-        vm = self.middleware.call_sync('vm._get_instance', id)
         if vm['name'] in self.vms:
             self.vms.pop(vm['name']).undefine_domain()
         else:
             VMSupervisor(vm, self.libvirt_connection, self.middleware).undefine_domain()
 
-        return self.middleware.call_sync('datastore.delete', 'vm.vm', id)
+        result = self.middleware.call_sync('datastore.delete', 'vm.vm', id)
+        if not self.middleware.call_sync('vm.query'):
+            self.close_libvirt_connection()
```

Comment:
Do we stop libvirtd when every VM is stopped?

Reply:
No, that would result in failure to managing them correctly. It's only closed if there are no more VM's left and the last one was deleted.

Reply:
I've meant when all VMs are stopped :)

1. Boot system with no VMs -> libvirtd is not running
2. Create VM - middleware will start libvirtd
3. Stop and delete VM - libvirtd will continue running

Is it like this?

Reply:
Yes it is. For step 3, we will close the hypervisor connection we have but libvirtd will continue running separately.

Reply:
Why?

Reply:
Why we still let libvirtd run in the background or why we close the connection ?

Reply:
Why we leave libvirtd running?

Reply:
We shouldn't, it provides little benefit when the vm is started again for opening a connection to the hypervisor but that's a fine tradeoff. Thank you for pointing it out